pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java

```diff
@@ -367,4 +410,98 @@ protected void onAckTimeoutSend(Set<MessageId> messageIds) {
             interceptors. onAckTimeoutSend(this, messageIds);
         }
     }
+
+    protected boolean canEnqueueMessage(Message<T> message) {
+        // Default behavior, can be overridden in subclasses
+        return true;
+    }
+
+    protected boolean enqueueMessageAndCheckBatchReceive(Message<T> message) {
+        if (canEnqueueMessage(message)) {
+            incomingMessages.add(message);
+            INCOMING_MESSAGES_SIZE_UPDATER.addAndGet(this, message.getData().length);
+        }
+        return hasEnoughMessagesForBatchReceive();
+    }
+
+    protected boolean hasEnoughMessagesForBatchReceive() {
+        if (batchReceivePolicy.getMaxNumMessages() <= 0 && batchReceivePolicy.getMaxNumMessages() <= 0) {
+            return false;
+        }
+        return (batchReceivePolicy.getMaxNumMessages() > 0 && incomingMessages.size() >= batchReceivePolicy.getMaxNumMessages())
+                || (batchReceivePolicy.getMaxNumBytes() > 0 && INCOMING_MESSAGES_SIZE_UPDATER.get(this) >= batchReceivePolicy.getMaxNumBytes());
+    }
+
+    private PulsarClientException verifyConsumerState() {
```

Comment:
returning an exception is not a good pattern in Java. I would suggest just throwing PulsarClientException. The caller can catch it and handle it properly.

Reply:
fix it